### ISSUE TITLE
False positive in "Squiz.PHP.DisallowSizeFunctionsInLoops" for "do ... while" loop

### ISSUE THREAD

**Issue by aik099** [squizlabs/PHP_CodeSniffer#764]
The `Squiz.PHP.DisallowSizeFunctionsInLoops` sniff checks, that `count(...)` function should not be used in following constructs:

```php
for ($i = 0; $i < count($records); $i++) {
    ...
}

while ( $i < count($records) ) {
    ...
    $i++;
}
```
This is perfectly logical. Unfortunately sniff also reacts to the `do ... while` loop:

```php
do {
    $records = $this->getRecordsSomehow();

} while ( count($records) > 0 );
```

In this case whatever or not to execute next iteration is based on external criteria, that can't be calculated upfront. This is usually the case with `do ... while` loops.

I propose to avoid reporting the `Squiz.PHP.DisallowSizeFunctionsInLoops.Found` error for `do ... while` constructs.

**Comment by gsherwood** [comment]
I don't think there is much difference between a regular `while` and a `do...while` when it comes to this check. The `do...while` is always going to run that first iteration, allowing you to create your records and set a count if you need to. This is the same process you'd go through in a regular `while` loop when you already have your array of records.

The check is there to ensure you aren't doing the `count()` unless you need to. If your `do...while` isn't actually changing the records array on every loop, then it shouldn't be counting on every loop. The sniff has no idea, so it throws up the error.

In a similar way, you might have a regular `while` loop that actually does change the records array and does need to count on every iteration. Banning the count in this case while allowing the `do...while` doesn't make any sense.

**Comment by aik099** [comment]
Maybe we can check if a counted variable is assigned inside a loop and then don't report that error?

Even if code would look like this we still count an assignment:

```php
do {
    $records['some_index'] = $this->getRecordsSomehow();

} while ( count($records['some_other_index']) > 0 );
```

**Comment by gsherwood** [comment]
> Maybe we can check if a counted variable is assigned inside a loop and then don't report that error?

That's a lot of effort, and still doesn't really get to the core reason why this sniff exists. It exists to say "don't run a counting function on each iteration unless you need to" and it is impossible to determine if the counting function is needed without checking if the counted variable has changed. You might only increment during some iterations based on a condition, but you need to re-count when you do that and don't need to re-count when you don't meet that condition.

What this sniff wants is for you to review your code and see if that count is really needed. If you want to code more defensively, don't include counting functions in the while condition at all and just do your counting in the iteration when required. That's specifically what this sniff is asking you to do and is how I code within the Squiz standard.

**Comment by aik099** [comment]
Then this sniff doesn't care about performance by suggesting moving out `count(...)` call to a variable, but just doesn't like `count(...)` to be present in loop condition in any case?

**Comment by gsherwood** [comment]
It does care about performance. It wants you to write code so that it is clear that the `count()` call is only ever used when needed instead of just defaulting to using it on every iteration.

Moving it out into a variable isn't going to affect performance if you need to count each time. It's the same number of `count()` calls if you do the first count before the `while` loop and then at the end of each iteration as it is if you do it in the `while` condition itself.

**Comment by aik099** [comment]
I personally don't like introducing new variable if it's just use once right on the next line (this is what needs to be done to satisfy this sniff). I guess in this case I'll just ignore that message (from sniff) because code like this isn't what I'm after for sure:

```php
do {
    $records = $this->getRecordsSomehow();
    // do something with records
    $recordCount = count($records);
} while ( $recordCount > 0 );
```
